python/pyarrow/_dataset_parquet_encryption.pyx

```diff
@@ -0,0 +1,170 @@
+
+    cdef shared_ptr[CParquetEncryptionConfig] unwrap(self):
+        return self.c_config
+
+
+cdef class ParquetDecryptionConfig(_Weakrefable):
+    """
+    Core configuration class encapsulating parameters for high-level decryption
+    within the Parquet framework.
+
+    ParquetDecryptionConfig is designed to pass decryption-related parameters to
+    the appropriate decryption components within the Parquet library. It holds references to
+    objects that define the decryption strategy, Key Management Service (KMS) configuration,
+    and specific decryption configurations for reading encrypted Parquet data.
+
+    Parameters
+    ----------
+    crypto_factory : pyarrow.parquet.encryption.CryptoFactory
+        Shared pointer to a `CryptoFactory` object, pivotal in creating cryptographic
+        components for the decryption process.
+    kms_connection_config : pyarrow.parquet.encryption.KmsConnectionConfig
+        Shared pointer to a `KmsConnectionConfig` object, containing parameters necessary
+        for connecting to a Key Management Service (KMS) during decryption.
+    decryption_config : pyarrow.parquet.encryption.DecryptionConfiguration
+        Shared pointer to a `DecryptionConfiguration` object, specifying decryption settings
+        for reading encrypted Parquet data.
+
+    Raises
+    ------
+    ValueError
+        Raised if `decryption_config` is None.
+    """
+
+    cdef:
+        shared_ptr[CParquetDecryptionConfig] c_config
+
+    # Avoid mistakingly creating attributes
+    __slots__ = ()
+
+    def __cinit__(self, CryptoFactory crypto_factory, KmsConnectionConfig kms_connection_config,
+                    DecryptionConfiguration decryption_config):
+
+        cdef shared_ptr[CDecryptionConfiguration] c_decryption_config
+
+        if decryption_config is None:
+            raise ValueError(
+                "decryption_config cannot be None")
+
+        self.c_config.reset(new CParquetDecryptionConfig())
+
+        c_decryption_config = pyarrow_unwrap_decryptionconfig(
+            decryption_config)
+
+        self.c_config.get().crypto_factory = pyarrow_unwrap_cryptofactory(crypto_factory)
+        self.c_config.get().kms_connection_config = pyarrow_unwrap_kmsconnectionconfig(
+            kms_connection_config)
+        self.c_config.get().decryption_config = c_decryption_config
+
+    @staticmethod
+    cdef wrap(shared_ptr[CParquetDecryptionConfig] c_config):
+        cdef ParquetDecryptionConfig python_config = ParquetDecryptionConfig.__new__(ParquetDecryptionConfig)
+        python_config.c_config = c_config
+        return python_config
+
+    cdef shared_ptr[CParquetDecryptionConfig] unwrap(self):
+        return self.c_config
+
+
+def set_encryption_config(ParquetFileWriteOptions opts, config):
+    cdef shared_ptr[CParquetEncryptionConfig] c_config
+    if not isinstance(config, ParquetEncryptionConfig):
+        raise ValueError("config must be a ParquetEncryptionConfig")
```

Comment:
If you can live with a `TypeError` instead of a `ValueError`, you can just type the argument instead of spelling out an additional type check.
```suggestion
def set_encryption_config(ParquetFileWriteOptions opts not None, ParquetEncryptionConfig config not None):
    cdef shared_ptr[CParquetEncryptionConfig] c_config
```
BTW, if you use Python argument annotation syntax, then (unless you use `Optional[]`) `None` will be rejected automatically. In C-style arguments, `None` is allowed (for historical reasons).

Reply:
> If you can live with a `TypeError` instead of a `ValueError`, you can just type the argument instead of spelling out an additional type check.

Ah, yes, and a TypeError is actually more correct

> BTW, if you use Python argument annotation syntax, then (unless you use `Optional[]`) `None` will be rejected automatically. In C-style arguments, `None` is allowed (for historical reasons).

Yes, we use `not None` in several places. Here the code has already checked beforehand that the value was not None (and the first argument is always `self`, so can't be None) and the function is not directly exposed to users, but also does no harm adding it to be sure.

We historically have all pyx files, and not python files with annotation syntax (and I also assume that for most files we wouldn't be able to use it anyway, because of our complex C++ wrapping cython usage, although I am probably not up to date with what the python syntax nowadays is able to do)